src/lib/message/ExchangeContext.cpp

```diff
@@ -0,0 +1,1664 @@
+        }
+        else
+        {
+            VerifyOrExit(mMsgProtocolVersion == kChipMessageVersion_V2, err = CHIP_ERROR_WRONG_MSG_VERSION_FOR_EXCHANGE);
+        }
+    }
+#endif
+    if (kChipMessageVersion_Unspecified == mMsgProtocolVersion)
+    {
+        mMsgProtocolVersion = msgInfo->MessageVersion = kChipMessageVersion_V1;
+    }
+    else
+    {
+        msgInfo->MessageVersion = mMsgProtocolVersion;
+    }
+
+    // Prevent sending if the context was associated with a connection that is now closed.
+    VerifyOrExit(!IsConnectionClosed(), err = CHIP_ERROR_NOT_CONNECTED);
+
+    // TODO: implement support for retransmissions.
+
+    // flag validation
+    if (sendFlags & kSendFlag_RetransmissionTrickle)
+    {
+        // We do not allow RMP to be used when Trickle retransmission is requested
+        if (sendFlags & kSendFlag_RequestAck)
+        {
+            ExitNow(err = CHIP_ERROR_INVALID_ARGUMENT);
+        }
+
+        // We do not support trickle retrasnmissions over
+        // connection-oriented exchanges
+        VerifyOrExit(Con == NULL, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+        if (0 == RetransInterval)
+        { // we're not retransmitting, do not hold onto the buffer
+            sendFlags &= ~kSendFlag_RetainBuffer;
+        }
+        else
+        {
+            sendFlags |= kSendFlag_RetainBuffer;
+            msg = msgBuf;
+        }
+    }
+
+    // Add the exchange header to the message buffer.
+    ChipExchangeHeader exchangeHeader;
+    memset(&exchangeHeader, 0, sizeof(exchangeHeader));
+
+    err = EncodeExchHeader(&exchangeHeader, profileId, msgType, msgBuf, sendFlags);
+    SuccessOrExit(err);
+
+    // If a response message is expected...
+    if ((sendFlags & kSendFlag_ExpectResponse) != 0)
+    {
+        // Only one 'response expected' message can be outstanding at a time.
+        VerifyOrExit(!IsResponseExpected(), err = CHIP_ERROR_INCORRECT_STATE);
+
+        SetResponseExpected(true);
+
+        // Arm the response timer if a timeout has been specified.
+        if (ResponseTimeout > 0)
+        {
+            err = StartResponseTimer();
+            SuccessOrExit(err);
+        }
+    }
+
+    // Fill in appropriate message header flags
+    if (sendFlags & kSendFlag_DelaySend)
+        msgInfo->Flags |= kChipMessageFlag_DelaySend;
+    // FIXME: RS: possibly unnecessary, should addref instead
+    if (sendFlags & kSendFlag_RetainBuffer)
+        msgInfo->Flags |= kChipMessageFlag_RetainBuffer;
+    if (sendFlags & kSendFlag_AlreadyEncoded)
+        msgInfo->Flags |= kChipMessageFlag_MessageEncoded;
+    if (sendFlags & kSendFlag_ReuseMessageId)
+        msgInfo->Flags |= kChipMessageFlag_ReuseMessageId;
+    if (sendFlags & kSendFlag_ReuseSourceId)
+        msgInfo->Flags |= kChipMessageFlag_ReuseSourceId;
+    if (sendFlags & kSendFlag_DefaultMulticastSourceAddress)
+        msgInfo->Flags |= kChipMessageFlag_DefaultMulticastSourceAddress;
```

Comment:
For all of these flags, are all these features needed in CHIP? I looked in the header, and it's not clear to me what most of these do...